### ISSUE TITLE
New Rule: Detect potential undefined property

### ISSUE THREAD

**Issue by erfanio** [eslint/eslint#8895]
**Please describe what the rule should do:**
Detect if an undefined object property is being used. This might not always be able to verify the structure of the object, but it would be invaluable to detect problems down the road before they bubble up (this is especially annoying since javascript silently fails and just returns an `undefined`)

**What category of rule is this?**

- [ ]  Enforces code style
- [x] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
```js
const foo = {
  bar: 'bar',
  hi: 'hi'
};
console.log(foo.baz); // Potential problem!
```

```js
import foo from './foo';
console.log(foo.baz); // Not sure
```

```js
const foo = {
  bar: 'bar',
  hi: 'hi'
};
console.log(foo.bar); // No problem
```

**Why should this rule be included in ESLint (instead of a plugin)?**
This would be useful for all developers. Everyone uses objects, but most importantly it would be valuable in places where you define a bunch of constants in an object (to be easily importable in other parts of the program).
Some people have encountered this issue and fixed it in their own community but easier detection would be super valuable. (for example [in redux](https://github.com/reactjs/redux/issues/541))

**Comment by erfanio** [comment]
After writing this I realised that this probably too much to ask from a linter and would be the job of a type checking system.
Feel free to close this issue

**Comment by nzakas** [comment]
Yeah, while we might be able to hack something together that kind of works, it's a lot harder than it looks. You really do need a type system to get the functionality you want.
